app/pb/communities.proto

```diff
@@ -0,0 +1,81 @@
+syntax = "proto3";
+
+package org.couchers.api.communities;
+
+import "google/protobuf/timestamp.proto";
+
+service Communities {
+  // This is a secure service: a user needs to be authenticated and not jailed to call functions here, refer to auth.proto and jail.proto
+
+  // Discussion threads are organized as a tree with maximum three levels depth:
+  // thread1
+  // +- comment1
+  // |  +- reply1
+  // |  +- reply2
+  // +- comment2
+  //    +- reply3
+  //    +- reply4
+  // thread2
+  // +- comment3
+  // |  +- reply5
+  // |  +- reply6
+  // +- comment4
+  //    +- reply7
+  //    +- reply8
+
+  // The same API is used regardless of level. All posts are using the same id namespace.
+
+  // Get non-deleted posts belonging to a thread, sorted in reverse time order.
+  // Pagination is supported by the posted_before_thread_id and number fields.
+  rpc ReadThread(ReadThreadReq) returns (ReadThreadRes);
+
+  // Post a new comment on a thread. The post can not be changed or deleted after posting.
+  // Throws INVALID_ARGUMENT if thread_id corresponds to a nesting level 3 reply, since no futher nesting is allowed.
+  rpc PostComment(PostCommentReq) returns (PostCommentRes);
+}
+
+message ReadThreadReq {
+  int64 thread_id = 1;
+  // Only get comments posted before the comment having this thread_id. Pass 0 to get all.
+  int64 posted_before_thread_id = 2;
+  // Maximum number of comments to return.
+  uint32 number = 3;
+}
+
+message ReadThreadRes {
+  // sorted in inverse time order, latest comment first.
+  repeated Comment comments = 1;
+
+  // number of comments posted earlier than the ones returned in `comments`
+  // zero if all comments has been returned to the client.
+  uint32 num_earlier_comments = 2;
+}
```

Comment:
Would it be good for this to return a "page token", i.e. the value to send to `posted_before_thread_id` to get an earlier list of messages? (I know this can be derived from `comments`). I'm not sure, what do you think?

Reply:
Yes, you pass the thread_id of the last item you got in the previous call :)

Reply:
That works!